retrospective-forecasting/main.py

```diff
@@ -181,49 +167,72 @@
-            model_name = forecast_path.stem.split("_")[1]
-            print_message(
-                f"Evaluating {model_name} model using {metric_name}...", end=""
+    for forecast_path in forecast_dir.glob("forecasts_*.parquet"):
+        model_name = forecast_path.stem.split("_")[1]
+        forecast = pl.scan_parquet(forecast_path)
+
+        for evaluator_config in config["evaluation"]["metrics"]:
+            if isinstance(evaluator_config, dict):
+                assert (
+                    len(evaluator_config) == 1
+                ), "Evaluator config is formatted incorrectly."
+
+                evaluator_config = list(evaluator_config.items())
+                evaluator_name = evaluator_config[0][0]
+                evaluator_args = {
+                    k: v for d in evaluator_config[0][1] for k, v in d.items()
+                }
+
+            else:
+                evaluator_name = evaluator_config
+                evaluator_args = {}
+
+            evaluator = getattr(linmod.eval, evaluator_name)(
+                samples=forecast,
+                data=eval_data.lazy(),
+                **evaluator_args,
             )
 
-            forecast = pl.read_parquet(forecast_path)
+            for metric_name, metric_function in vars(type(evaluator)).items():
+                if metric_name.startswith("_"):
+                    continue
 
-            forecast = forecast.lazy()
-            scores.append(
-                (
-                    metric_name,
-                    model_name,
-                    metric_function(forecast, eval_data.lazy()),
+                print_message(
+                    (
+                        f"Evaluating {model_name} model using "
+                        f"{evaluator_name}.{metric_name}..."
+                    ),
+                    end="",
                 )
-            )
 
-            if (
-                metric_name == config["evaluation"]["metrics"][0]
-            ):  # One plot per model
-                data_path = config["data"]["save_file"]["eval"]
-                png_path = (
-                    eval_dir / "visualizations" / f"eval_{model_name}.png"
-                )
-                plot_script.write(
+                scores.append(
                     (
-                        "python3 -m linmod.visualize "
-                        f"-f {forecast_path} "
-                        f"-d {data_path} "
-                        f"-p {png_path} "
-                        "-t eval\n"
+                        f"{evaluator_name}.{metric_name}",
+                        model_name,
+                        metric_function(evaluator),
                     )
                 )
 
-            print_message(" done.")
+                print_message(" done.")
+
+        data_path = config["data"]["save_file"]["eval"]
+        png_path = eval_dir / "visualizations" / f"eval_{model_name}.png"
+        plot_script.write(
```

Comment:
This is giving me serious (bad) chills

You're writing python code programmatically?

This is telling me that we really want a snakefile, that can know that we want to do a number of tasks for every combination of model, region, forecast date(?)

Reply:
This was more of a quick patch for a memory leak that Andy pushed in #66. While we both agree that it definitely needs to go, it also feels not urgent given that I also think `retrospective-forecasting/main.py` has grown to be a bit unwieldy and should probably be reworked itself too

Reply:
Given that this is already in `main` and is probably a fairly large task, I’d say it’ll be its own PR, though?